### PR TITLE
Initial write up of June scale test

### PR DESCRIPTION
**Pull request by pricem14pc**
Documents the scale testing for June 2020, observations and recommendations.

Does it make sense?
Any further recommendations?

### CODE REVIEW

File: doc/performance-testing/stress-test-june-2020.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+## Recommendations
+
+- Runner is only utilising 49% of requested available vCPU. Test different gunicorn configurations and investigate other WSGI or event loop architectures
+- Resolve/understand the 401 errors we have seen across this test
+- Redis memorystore CPU 100%, test with increased spec to see if additional resource allows us to scale runner further
+- Speak to GCP Redis engineers to understand this behaviour and if our useage is appropriate
+- Tune and retest the kubernetes autoscaling and configuration (e.g is the 50% `target_cpu_utilization_percentage` appropriate)
+- Update Grafana dashboard to add vCPU and Launch / Submission counts
+- Update summary script to aggregate scale test results per folder more appropriately (inc. Total Requests, Failure %, Max response time)
```
Comment (MebinAbraham) [inline doc/performance-testing/stress-test-june-2020.md:71]: 👍

File: doc/performance-testing/stress-test-june-2020.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+- The average submission rate in the 90 instances test was 135 responses per second, which is 486,000 responses per hour (assuming requests remain stable over an hour)
+- 30,000 rps with 765 used vCPU is reasonable. This equates to 14.32 rps per core (as a GCP resource where the 4th core is effectively unusable)
+
+## Recommendations
+
+- Runner is only utilising 49% of requested available vCPU. Test different gunicorn configurations and investigate other WSGI or event loop architectures
+- Resolve/understand the 401 errors we have seen across this test
+- Redis memorystore CPU 100%, test with increased spec to see if additional resource allows us to scale runner further
+- Speak to GCP Redis engineers to understand this behaviour and if our useage is appropriate
+- Tune and retest the kubernetes autoscaling and configuration (e.g is the 50% `target_cpu_utilization_percentage` appropriate)
```
Comment (MebinAbraham) [inline doc/performance-testing/stress-test-june-2020.md:69]:
```suggestion
- Tune and retest the Kubernetes autoscaling and configuration (e.g. is the 50% `target_cpu_utilization_percentage` appropriate)
```

File: doc/performance-testing/stress-test-june-2020.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+- 401 errors x 10,038 throughout (distributed across all pods)
+        
+        8,353 x flask_wtf.csrf.CSRFError: 400 Bad Request: The CSRF tokens do not match."
+        2,028 x flask_wtf.csrf.CSRFError: 400 Bad Request: The CSRF session token is missing."
+
+- 502 errors x 1,075 at 110 load injectors reported by load_balancer as `backend_timeout`
+
+## Observations
+
+- The run of 110 instances saw significant increase in response times resulting in request timeouts. This coincided with Redis reaching 100% CPU. At this point the load balancer began seeing the 502 `backend_timeout` errors
```
Comment (MebinAbraham) [inline doc/performance-testing/stress-test-june-2020.md:59]:
```suggestion
- The run of 110 benchmark instances saw a significant increase in response times resulting in request timeouts. This coincided with Redis reaching 100% CPU. At this point, the load balancer began seeing the 502 `backend_timeout` errors
```

File: doc/performance-testing/stress-test-june-2020.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
+# Scale Test June 2020
```
[the remark this answers is not in the document]
Comment (pricem14pc) [reply doc/performance-testing/stress-test-june-2020.md:1]: Not sure it needs a folder per test at this stage. I've prefixed the file names instead.

**Pull request merged**